### ISSUE TITLE
[Task 18] Format-aware reward router

### ISSUE THREAD

**Issue by aayc123** [redai-infra/Relax#97]
Task 18

## 1. Repository, Baseline Commit, and Development Branch

- Repository: https://github.com/redai-infra/Relax
- Baseline branch: `main`
- Baseline commit: `039ce876d25540adad847d4223b4de4722d8f425`
- Development branch: `feat/task18-format-aware-reward-router`

Development and testing will be based on the commit above. The Pull Request will not contain changes unrelated to Task 18.

## 2. Runtime Environment

| Item         | Configuration                                                |
| ------------ | ------------------------------------------------------------ |
| Image        | `ghcr.io/redai-infra/relaxrl:dev-20260715-8325919e`          |
| Image digest | `sha256:3fa8ce578acda6c829b83016bde42c38fa892681e4f36ca330f545616fe578e2` |
| Python       | 3.12.3                                                       |
| CUDA         | 12.9                                                         |
| PyTorch      | 2.11.0+cu129                                                 |
| Ray          | 2.56.0                                                       |
| SGLang       | 0.5.12.post1                                                 |
| GPU          | 1 × NVIDIA H800 PCIe 80 GB                                   |
| CPU          | 2 × Intel Xeon Platinum 8558, 96 cores / 192 threads         |
| Memory       | 1.0 TiB                                                      |
| OS           | Ubuntu 22.04.5 LTS                                           |

## 3. Launch Script and Complete Commands

This task does not add a training launch script or introduce or modify CLI arguments. The existing `--rm-type` argument remains the explicit global reward type and preserves the behavior of existing single-task training scripts. A valid per-sample `metadata["rm_type"]` takes precedence over it, while label-based inference is only used when neither metadata nor `--rm-type` resolves to a registered reward type.

Start the test image:

```bash
docker run --rm -it \
  --network host \
  --shm-size=64g \
  -v /data/zn/Relax:/workspace/Relax \
  ghcr.io/redai-infra/relaxrl:dev-20260715-8325919e \
  bash
```

Run targeted unit tests:

```bash
cd /workspace/Relax
python -m pytest -v tests/engine/rewards/test_reward_worker.py
```

If a separate router test file is added, run it as well:

```bash
cd /workspace/Relax
python -m pytest -v tests/engine/rewards/test_reward_router.py
```

Run the reward regression suite:

```bash
cd /workspace/Relax
python -m pytest -v tests/engine/rewards/
```

Run code checks before submission:

```bash
cd /workspace/Relax
pre-commit run --all-files
```

- New parameters: none
- Modified parameters: none

### Related Files

The main files are expected to be:

- `relax/engine/rewards/__init__.py`
- `tests/engine/rewards/test_reward_worker.py`

Depending on the final code structure, the implementation may also:

- add a separate reward router/registry module;
- add `tests/engine/rewards/test_reward_router.py`;
- make small necessary changes to the math and multiple-choice reward modules.

### Core Functions

The planned routing precedence is:

```text
valid metadata["rm_type"]
→ valid explicit --rm-type
→ exactly one matching label matcher
→ return a reward-key-aware zero and emit a warning
```

The planned changes are:

1. Introduce a unified reward registry containing reward handlers and optional label matchers.
2. Prefer a valid per-sample `metadata["rm_type"]`.
3. If metadata does not resolve to a registered reward type, preserve the existing explicit `--rm-type` behavior.
4. Only when neither metadata nor explicit `--rm-type` resolves successfully, use registered label matchers for automatic format inference.
5. If exactly one label matcher succeeds, use the corresponding reward.
6. For unknown metadata, unknown explicit types, conflicting label matches, or unresolved routes, emit a readable warning.
7. If no valid route is available, return `0.0`, or `{reward_key: 0.0}` when a reward key is configured.
8. Adding a reward should require one registry entry without modifying generic router branches.
9. Preserve compatibility with existing single-task `--rm-type` workflows.

The explicit `--rm-type` remains authoritative for existing training scripts. This prevents numeric labels used by DAPO, GenRM, OpenR1MM, or other reward schemas from being incorrectly inferred as the scalar `math` reward.

For mixed-task batches, per-sample `metadata["rm_type"]` is the preferred explicit routing mechanism. Label inference provides a fallback for samples where no valid metadata or global reward type is available.

The initial label matchers will only cover the math and multiple-choice formats required for acceptance. This task will not implement a general natural-language classifier.

### Non-goals

This task will not:

- modify GRPO, PPO, DAPO, or other RL algorithms;
- modify Controller, Ray Serve, or rollout scheduling;
- modify the public `Sample.label` or `Sample.metadata` structures;
- add third-party dependencies;
- change training dataset formats;
- change existing reward scoring formulas;
- bulk-add `rm_type` to existing datasets;
- delete or rename existing public APIs;
- optimize training, throughput, GPU memory, or convergence;
- change the existing single-task `--rm-type` workflow.

### Baseline Data and Behavior

At commit `039ce876d25540adad847d4223b4de4722d8f425`:

- global `--rm-type` is supported;
- `metadata["rm_type"]` can override the global type;
- `label` is primarily used as the ground-truth answer;
- no unified label-matcher registry exists;
- mixed-task batches primarily depend on samples providing the correct `metadata["rm_type"]`.

Unit tests will use minimal in-memory `Sample` objects without downloading models or training datasets. Coverage will include:

- math samples;
- multiple-choice samples;
- a mixed math and multiple-choice batch;
- valid and unknown per-sample metadata types;
- explicit `--rm-type` taking precedence over label inference;
- label inference when no valid explicit type is available;
- missing or conflicting label matches;
- configured and missing fallback routes;
- reward-key-aware zero handling;
- registration of a test reward;
- compatibility with existing explicit `--rm-type`.

### Acceptance Metrics

| Requirement            | Measurement                                                     | Target                                                          |
| ---------------------- | --------------------------------------------------------------- | --------------------------------------------------------------- |
| Mixed-batch routing    | Check reward selection for each sample in the same batch        | All math and multiple-choice samples are routed correctly       |
| Metadata routing       | Valid `metadata["rm_type"]`                                     | Uses the per-sample specified reward                            |
| Explicit routing       | Valid explicit `--rm-type` with no valid metadata override      | Uses the explicit reward before label inference                 |
| Label routing          | No valid metadata or explicit type, with one matching label     | Uses the uniquely inferred reward                               |
| Fallback               | Unknown metadata with a valid explicit `--rm-type`              | Uses the configured explicit reward                             |
| Conflict handling      | Multiple label matchers succeed and no valid explicit type      | Returns zero and records a warning                              |
| Zero-score handling    | No valid metadata, explicit type, or unique label match         | Returns `0.0` or `{reward_key: 0.0}`                            |
| Warning                | Unknown, conflicting, or unresolved routing information         | Emits a readable warning containing the reason                  |
| Registry               | Register a test reward                                          | One registration entry with no generic router branch changes    |
| Backward compatibility | Existing explicit `--rm-type` cases                             | Preserve their original reward selection and output schema      |
| Reward regression      | `tests/engine/rewards/`                                         | All pass                                                        |

Tests will use deterministic inputs without random sampling. The primary acceptance criterion is that the targeted unit tests and reward regression suite pass completely.

## 6. Expected Delivery Method

Delivery method: GitHub Issue + Pull Request.

The Pull Request is expected to include:

- a one-line reward registry;
- per-sample routing based on metadata and label formats;
- fallback, zero-score, and warning handling;
- backward compatibility with explicit `--rm-type`;
- registry, routing, fallback, conflict, and mixed-batch tests;
- pytest and code-check results.
